ci/verify_examples.sh

```diff
@@ -1,43 +1,524 @@
-#!/bin/bash
+#!/bin/bash -E
 
-set -e
+FAILED=()
+SRCDIR="${SRCDIR:-$(pwd)}"
+EXCLUDED_BUILD_CONFIGS=${EXCLUDED_BUILD_CONFIGS:-"^./jaeger-native-tracing|docker-compose"}
 
-verify() {
-  echo $1
-  CONTAINER_ID="$(docker ps -aqf name=$1)"
-  if [ "false" == "$(docker inspect -f {{.State.Running}} ${CONTAINER_ID})" ]
-  then
-    echo "error: $1 not running"
+
+trap_errors () {
+    local frame=0 COMMAND LINE SUB FILE
+    if [ -n "$example_test" ]; then
+        COMMAND=" (${example_test})"
+    fi
+    set +v
+    while read -r LINE SUB FILE < <(caller "$frame"); do
+        if [ "$frame" -ne "0" ]; then
+            FAILED+=("  > ${SUB}@ ${FILE} :${LINE}")
+        else
+            FAILED+=("${SUB}@ ${FILE} :${LINE}${COMMAND}")
+        fi
+        ((frame++))
+    done
+    set -v
+}
+
+trap trap_errors ERR
+trap exit 1 INT
+
+run_log () {
+    local name
+    name="$1"
+    shift
+    echo -e "\n> [${name}] ${*}"
+}
+
+show_user_env () {
+    run_log "$(whoami)" "User env"
+    id
+    echo "umask = $(umask)"
+    echo "pwd = $(pwd)"
+}
+
+get_path () {
+    printf "%s/examples/%s" "$SRCDIR" "$1"
+}
+
+bring_up_example_stack () {
+    local args name path snooze
+    args=("${@}")
+    name="$1"
+    path="$2"
+    snooze="${3:-0}"
+    cd "$path" || return 1
+    run_log "$name" "Pull the images"
+    docker-compose pull || return 1
+    echo
+    run_log "$name" "Bring up services"
+    docker-compose up --build -d "${args[@]:3}" || return 1
+    if [ "$snooze" -ne "0" ]; then
+        run_log "$name" "Snooze for ${snooze} while ${name} gets started"
+        sleep "$snooze"
+    fi
+    docker-compose ps
+    docker-compose logs
+}
+
+bring_up_example () {
+    local name paths
+    name="$1"
+    read -ra paths <<< "$(echo "$2" | tr ',' ' ')"
+    shift 2
+    for path in "${paths[@]}"; do
+        bring_up_example_stack "$name" "$(get_path "$path")" "$@"
+    done
+}
+
+cleanup_stack () {
+    local name path
+    name="$1"
+    path="$2"
+    run_log "$name" "Cleanup: $path"
+    cd "$path" || return 1
+    docker-compose down
+    docker system prune -f
+}
+
+cleanup () {
+    local name paths
+    name="$1"
+    read -ra paths <<< "$(echo "$2" | tr ',' ' ')"
+    for path in "${paths[@]}"; do
+        cleanup_stack "$name" "$(get_path "$path")"
+    done
+}
+
+run_example_cors () {
```

Comment:
Shall we have each of these function into a separate script so they can be run individually? How do I debug if one of the example test failed? And verify_examples can just glob in sample directory to run them?

Reply:
> Shall we have each of these function into a separate script so they can be run individually? 

not sure. its defo an issue we need to address - i was un/commenting as i went through  the tests

im wondering if adding an argument - eg `ci/verify_examples.sh jaeger*` - might be a good idea - irrespective of whether all of the tests are actually in the file.

i did think about putting `verify.sh` in each of the example folders - to be more explicit for example creators - but then i thought it was distracting for example users

> How do I debug if one of the example test failed? 

you get a bash traceback with the line numbers - this actually works pretty well i think

> And verify_examples can just glob in sample directory to run them?

not sure if i understand exactly this point.

i think we do want to glob in the actual examples folder to work out what tests *should* exist to enforce that examples need a test

somewhat tangentially i had some ideas to make this script a little more concise/readable:

- remove the local `name` in each test - there is already a usable var in scope
- add some  helper functions like `responds_with`  `responds_with_header` `responds_without_header` - this would make the curl tests more explicit as to what they were testing i think

ill update with  these 2 changes shortly - unless you think its a bad idea

Reply:
> > Shall we have each of these function into a separate script so they can be run individually?
> 
> not sure. its defo an issue we need to address - i was un/commenting as i went through the tests
> 
> im wondering if adding an argument - eg `ci/verify_examples.sh jaeger*` - might be a good idea - irrespective of whether all of the tests are actually in the file.
> 
> i did think about putting `verify.sh` in each of the example folders - to be more explicit for example creators - but then i thought it was distracting for example users

How it would distracting? You can have verify_examples.sh run all verify.sh?
Then if I want to run only one example, I can go to that and run verify.sh there.